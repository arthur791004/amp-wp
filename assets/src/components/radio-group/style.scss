.radio-group__label {
	align-items: baseline;
	color: var(--amp-settings-color-muted);
	display: flex;
	flex-flow: row nowrap;

	&.radio-group__label--selected,
	&:hover {
		color: var(--amp-settings-color-black);
	}
}

.radio-group__title.radio-group__title {
	color: inherit;
	font-size: 16px;
}

.radio-group--stacked {

	.radio-group__label {
		padding: 15px 0 14px;

		& + .radio-group__label {
			border-top: 1px solid var(--amp-settings-color-border);
		}

		&:first-child {
			padding-top: 0;
		}

		&:last-child {
			padding-bottom: 0;
		}
	}

	.radio-group__input {
		margin-right: 20px;
	}

}

.radio-group--horizontal {
	display: flex;
	flex-flow: row nowrap;

	.radio-group__input {
		margin-right: 10px;
	}

	.radio-group__label {
		padding: 20px 0;

		& + .radio-group__label {
			margin-left: 25px;
		}
	}
}
