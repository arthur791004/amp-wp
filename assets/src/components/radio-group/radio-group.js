/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { useInstanceId } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import './style.scss';

/**
 * Radio group.
 *
 * @param {Object}   props              Component props.
 * @param {Array}    props.options      List of option values and titles.
 * @param {Function} props.onChange     Change handler.
 * @param {string}   props.selected     Currently selected option.
 * @param {boolean}  props.isHorizontal Flag indicating that the options should not be stacked.
 */
export function RadioGroup( { options = [], onChange, selected, isHorizontal } ) {
	const htmlIdPrefix = `radio-group-${ useInstanceId( RadioGroup ) }`;

	return (
		<form
			className={ classnames( 'radio-group', {
				'radio-group--stacked': ! isHorizontal,
				'radio-group--horizontal': isHorizontal,
			} ) }
		>
			{ options.map( ( { value, title } ) => (
				<label
					key={ value }
					className={ classnames( 'radio-group__label', {
						'radio-group__label--selected': isHorizontal,
					} ) }
					htmlFor={ `${ htmlIdPrefix }-${ value }` }
				>
					<div className="radio-group__input">
						<input
							type="radio"
							id={ `${ htmlIdPrefix }-${ value }` }
							checked={ selected === value }
							onChange={ () => onChange( value ) }
						/>
					</div>
					<h3 className="radio-group__title">
						{ title }
					</h3>
				</label>
			) ) }
		</form>
	);
}

RadioGroup.propTypes = {
	options: PropTypes.arrayOf(
		PropTypes.shape( {
			value: PropTypes.string,
			title: PropTypes.string,
		} ),
	),
	onChange: PropTypes.func,
	selected: PropTypes.string,
	isHorizontal: PropTypes.bool,
};
