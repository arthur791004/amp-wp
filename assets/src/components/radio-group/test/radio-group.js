
/**
 * External dependencies
 */
import { act } from 'react-dom/test-utils';
import { create } from 'react-test-renderer';

/**
 * WordPress dependencies
 */
import { render } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { RadioGroup } from '../radio-group';

let container;

const options = [
	{
		value: 'a',
		title: 'Foo',
	},
	{
		value: 'b',
		title: 'Bar',
	},
];

describe( 'RadioGroup', () => {
	beforeEach( () => {
		container = document.createElement( 'div' );
		document.body.appendChild( container );
	} );

	afterEach( () => {
		document.body.removeChild( container );
		container = null;
	} );

	it( 'matches the snapshot', () => {
		const wrapper = create( <RadioGroup options={ options } /> );

		expect( wrapper.toJSON() ).toMatchSnapshot();
	} );

	it( 'renders radio buttons in a form element', () => {
		act( () => {
			render(
				<RadioGroup options={ options } />,
				container,
			);
		} );

		expect( container.querySelector( 'form' ) ).not.toBeNull();
		expect( container.querySelector( 'form' ).className ).toContain( '--stacked' );
		expect( container.querySelectorAll( 'input[type="radio"]' ) ).toHaveLength( 2 );
		expect( container.querySelector( 'input[type="radio"]:checked' ) ).toBeNull();
		expect( container.querySelector( 'form' ).textContent ).toBe( 'FooBar' );
	} );

	it( 'renders label with a matching for attribute', () => {
		act( () => {
			render(
				<RadioGroup options={ options } />,
				container,
			);
		} );

		const forAttribute = container.querySelector( 'label:first-child' ).getAttribute( 'for' );
		expect( container.querySelector( 'label:first-child input' ).getAttribute( 'id' ) ).toBe( forAttribute );
	} );

	it( 'renders a horizontal group', () => {
		act( () => {
			render(
				<RadioGroup options={ options } isHorizontal={ true } />,
				container,
			);
		} );

		expect( container.querySelector( 'form' ).className ).toContain( '--horizontal' );
	} );

	it( 'renders an initially selected option', () => {
		act( () => {
			render(
				<RadioGroup options={ options } selected="b" />,
				container,
			);
		} );

		expect( container.querySelector( 'input[type="radio"]:checked' ) ).not.toBeNull();
		expect( container.querySelector( 'input[type="radio"]:checked' ).closest( 'label' ).textContent ).toBe( 'Bar' );
		expect( container.querySelector( 'input[type="radio"]:not(checked)' ) ).not.toBeNull();
	} );

	it( 'calls the handler function on selection change', () => {
		const handler = jest.fn();

		act( () => {
			render(
				<RadioGroup options={ options } onChange={ handler } />,
				container,
			);
		} );

		act(
			() => {
				container.querySelector( 'label' ).dispatchEvent( new global.MouseEvent( 'click' ) );
			},
		);

		expect( handler ).toHaveBeenCalledTimes( 1 );
		expect( handler ).toHaveBeenCalledWith( 'a' );
	} );
} );
