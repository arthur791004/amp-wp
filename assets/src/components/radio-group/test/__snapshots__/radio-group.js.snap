// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup matches the snapshot 1`] = `
<form
  className="radio-group radio-group--stacked"
>
  <label
    className="radio-group__label"
    htmlFor="radio-group-0-a"
  >
    <div
      className="radio-group__input"
    >
      <input
        checked={false}
        id="radio-group-0-a"
        onChange={[Function]}
        type="radio"
      />
    </div>
    <h3
      className="radio-group__title"
    >
      Foo
    </h3>
  </label>
  <label
    className="radio-group__label"
    htmlFor="radio-group-0-b"
  >
    <div
      className="radio-group__input"
    >
      <input
        checked={false}
        id="radio-group-0-b"
        onChange={[Function]}
        type="radio"
      />
    </div>
    <h3
      className="radio-group__title"
    >
      Bar
    </h3>
  </label>
</form>
`;
